### PR TITLE
Persisting the result of successful MP-to-Policy migration

### PR DESCRIPTION
**Pull request by joroaf**
Persisting the result of a successful MP-to-Policy in the Transport Zone tag, so on restart the agent:

- will not try to conduct a migration again
- will use only Policy API

### CODE REVIEW

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider.py [hunk 1]
```diff
@@ -413,3 +425,21 @@ def sanitize(self, slice: int) -> List[Tuple[str, Callable[[str], None]]]:
...
         """
+
+
+class MigrationTracker(object):
+    def __init__(self, provider: Provider) -> None:
+        self._migration_in_progress = False
+        self.mutex = "migration-tracking"
+        self.provider = provider
+
+    def set_migration_in_progress(self, in_progress: bool):
```
Comment (sven-rosenzweig) [inline networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider.py:436]: I would name it `set_migration_status`

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider.py [hunk 2]
```diff
@@ -413,3 +425,21 @@ def sanitize(self, slice: int) -> List[Tuple[str, Callable[[str], None]]]:
...
+        with LockManager.get_lock(self.mutex):
+            self._migration_in_progress = in_progress
+
+    def is_migration_in_progress(self):
+        with LockManager.get_lock(self.mutex):
+            return self._migration_in_progress
+
+    def store_migration_result(self, migration_scope: str, migration_result: str):
```
Comment (sven-rosenzweig) [inline networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider.py:444]: I would name it `persist_migration_status`

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py [hunk 3]
```diff
@@ -448,11 +447,24 @@ def __init__(self, payload: Payload = Payload(), zone_id: str = ""):
             self._ensure_default_l3_policy()
         self._setup_default_app_drop_logged_section()
 
-    def _load_zone(self):
+    def _get_tz(self) -> dict or None:
```
Comment (sven-rosenzweig) [inline networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:450]: _fetch_tagged_zone

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py [hunk 4]
```diff
@@ -448,11 +447,24 @@ def __init__(self, payload: Payload = Payload(), zone_id: str = ""):
...
+            if tz.get("display_name") == self.zone_name:
+                return tz
+        return None
+
+    def _load_zones(self):
         LOG.info("Looking for TransportZone with name %s.", self.zone_name)
-        for zone in self.client.get_all(path=API.SEARCH_QUERY, params={"query": API.SEARCH_Q_TRANSPORT_ZONES.format(self.zone_name)}):
-            if zone.get("display_name") == self.zone_name:
-                return zone.get("id")
+
```
Comment (sven-rosenzweig) [inline networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:458]: return self.tz()

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/realization.py [hunk 5]
```diff
@@ -34,19 +36,23 @@ def __init__(
...
         self.plcy_provider = plcy_provider
+        self.migration_tracker = provider.MigrationTracker(self.plcy_provider)
+
 
-        self.PAUSE_REALIZATION = False
         self.AGE = int(time.time())
 
         LOG.info("Detected NSX-T %s version.", self.mngr_provider.client.version)
 
+        self.mp_to_policy_completed = any([t for t in self.plcy_provider.zone_tags if t.get(
```
Comment (sven-rosenzweig) [inline networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/realization.py:46]: Should we move that to a separate method?

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/realization.py [hunk 6]
```diff
@@ -34,19 +36,23 @@ def __init__(
...
+
         # Enable MP-to-Policy migration if force_mp_to_policy=True
         # It is used as a flag for using Policy API completely or not
         # in case migration canceled or failed this flag will be False
         # TODO: After completing the transition to NSX Policy API (ONLY if successful!), deprecate this flag
-        self.USE_POLICY_API = self._check_mp2policy_support()
+        self.USE_POLICY_API = self.mp_to_policy_completed or self._check_mp2policy_support()
 
-        if self.USE_POLICY_API:
+        if self.USE_POLICY_API and not self.mp_to_policy_completed:
```
Comment (sven-rosenzweig) [inline networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/realization.py:55]: what about    if  self.mp_to_policy_completed: return to make it clear that migration has already happend.

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider.py (hunk 1 above)
Comment (joroaf) [reply networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider.py:436]: it is not about the status but the state, we just set a boolean flag which represents two states: "migration in progress", "migration not in progress"

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py [hunk 7]
```diff
@@ -448,11 +447,24 @@ def __init__(self, payload: Payload = Payload(), zone_id: str = ""):
             self._ensure_default_l3_policy()
         self._setup_default_app_drop_logged_section()
 
-    def _load_zone(self):
+    def _get_tz(self) -> dict or None:
+        for tz in self.client.get_all(path=API.SEARCH_QUERY, params={"query": API.SEARCH_Q_TRANSPORT_ZONES.format(self.zone_name)}):
+            if tz.get("display_name") == self.zone_name:
+                return tz
```
[the remark this answers is not in the document]
Comment (joroaf) [reply networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:453]: no, this method returns the entire TZ object, not only the tags

File: networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py (hunk 4 above)
Comment (joroaf) [reply networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:458]: this is concrete implementation of an abstract method which expects tuple as return

**Review by notandy — APPROVED** [review]

**Pull request merged**